### ISSUE TITLE
Cookies not fetched with WebAPI.getCookies()

### ISSUE THREAD

**Issue by Hugolarson** [JPro-one/JPro-tickets#52]
Setting a Cookie with java script and after try to read it with WebAPI.getWebAPI(stage).getCookies() does not fetch the Cookie until browser refresh.
It's possible to fetch it with java script though.

Don't know if i'm using getCookies() wrong.

It would be very nice if it's possible to also create Cookies with WebApi.

**Comment by FlorianKirmaier** [comment]
Sorry for the late answer. 

getCookies only returns the cookies which were existing during the start of the application.
A setCookies method is not trivial because there are a lot of topics about how it should work in detail.
The cookies API of JPro should probably be redesigned.
For that reason, I recommend for now to set cookies directly through javascript.

**Comment by Hugolarson** [comment]
Further comment on the subject.
JPro does not allow WebAPI.executeScriptWithReturn on Java FX application thread makes it cumbersome to use.
I think that reading and setting Cookies should be like reading and writing files since Cookies are the only way to store data on client side.
This is possible with Vaadin :)

**Comment by FlorianKirmaier** [comment]
I agree with you. This is a very important feature.

**Comment by FlorianKirmaier** [comment]
@Hugolarson 
We are adding a CookieAPI for the next version 2019.2.7.
It can already be tested with the version 2019.2.7-SNAPSHOT.

These are the methods of the WebAPI in the current SNAPSHOT version:
```
    /**
     * Sets a new cookie in the current browser session. Afterward, the map of getCookies() gets updated.
     * The cookie won't expire.
     */
    public abstract void setCookie(String key, String value);

    /**
     * Deletes a cookie in the current browser session. Afterward, the map of getCookies() gets updated.
     */
    public abstract void deleteCookie(String key);

    /**
     * @return the cookies of the browser page. Gets updated after setCookie() and deleteCookie().
     */
    public abstract ObservableMap<String,String> getCookies();
```

**Comment by samyc1201** [comment]
@FlorianKirmaier @Hugolarson 

I guess we can use cookies to manage user session in the jpro backend. But how do we intercept all the communications which go to server in order to validate user after they login? It is very troublesome if we add in such session validation everywhere where the UI event is triggered. Any advice?

**Comment by FlorianKirmaier** [comment]
You can basically do the same, you would do for a web or desktop application, or do I miss something?
There shouldn't really be something, which isn't possible.
You could add a periodic check, whether the session is still valid, or do it whenever the user does some kind of action.

**Comment by samyc1201** [comment]
@FlorianKirmaier Normally in web (Java EE), we have some sort of filter to verify each communication, so we don't need to add session validation everywhere in the business logic layer. Periodic check will decouple that but still there is chance to slip through during the gap of periodic check. 

In addition, I guess `WebAPI.getCookies()` will be triggered from backend, and the instruction flow to browser, retrieve the cookies, later on pass back to backend again, this is kind of extra round-trip, not sure would it impacts the app performance?

If we can do these below, that would be best as we can enforce the best security at backend:
1) allow to configure whether to pass back cookies in each websocket request,  for those who doesn't need, this feature can be disabled.
2) allow to apply some sort of listener to intercept websocket request at backend, so we can do the check prior we perform any Java backend action.

**Comment by FlorianKirmaier** [comment]
1. WebAPI.getCookies just remembers the cookies when the app hast started und combines it with the setCookie calls. 
It doesn't really have a performance impact. 

2. Yes, that would be useful! Not only for cookies, but also for other forms of verification.

Maybe in the future it's also possible to support the mechanism of the JavaEE world.

**Comment by Hugolarson** [comment]
How to set Cookie parameters such as Max-Age and Path?

**Comment by samyc1201** [comment]
@FlorianKirmaier Yes! It doesn't have to support whole JavaEE features, it is too huge, but at least the session management and access control can be in place, it will be very convenient for those app which involves login and require to serve multiple user at the same time. 

I found it a challenge to distinguish a login-ed user in JPro. Normally for a JavaFX desktop app, user id can be stored easily in local memory (there is no need to distinguish user in this case as it is separated JVM for each user). But JPro execute Java logic at remote JVM, and render UI at local browser, so I guess in this case I have to store login-ed user ID/token in browser side (via javascript API), and then rely on it to distinguish a user. Not very straightforward though.

**Comment by samyc1201** [comment]
@Hugolarson I was looking the same thing, I guess you can use [WebAPI](https://www.jpro.one/api/2019.1.2/com/jpro/webapi/WebAPI.html) to execute custom JavaScript. Something like `document.cookie = "doSomethingOnlyOnce=; expires=Thu, 01 Jan 1970 00:00:00 GMT";`

**Comment by Hugolarson** [comment]
@samyc1201 Well then the entire Cookie has to be set with JS with is not pretty at all.
I think JPro should have a Cookie class with all cookie parameters and set it with cookieMap.put(String name, Cookie c).

**Comment by Hugolarson** [comment]
Need to wake up this Cookie issue :)
How to set Cookie parameters such as Max-Age and Path with new JPRO Cookie API without JS calls?
